stablehlo/reference/Element.cpp

```diff
@@ -624,6 +613,161 @@ Element complex(const Element &e1, const Element &e2) {
+                  /*isUnsigned=*/isSupportedUnsignedIntegerType(type));
+    bool roundingErr;
+    value.convertToInteger(intVal, APFloat::rmTowardZero, &roundingErr);
+    return Element(type, intVal);
+  }
+  if (isSupportedFloatType(type)) {
+    bool roundingErr;
+    value.convert(type.cast<FloatType>().getFloatSemantics(),
+                  APFloat::rmNearestTiesToEven, &roundingErr);
+    return Element(type, value);
+  }
+  if (isSupportedComplexType(type)) {
+    const llvm::fltSemantics &fltSemantics = type.cast<ComplexType>()
+                                                 .getElementType()
+                                                 .cast<FloatType>()
+                                                 .getFloatSemantics();
+    bool roundingErr;
+    value.convert(fltSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    auto imagVal = APFloat(0.0);
+    imagVal.convert(fltSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    return Element(type, std::complex<APFloat>(value, imagVal));
+  }
+  report_fatal_error(invalidArgument("Unsupported element type: %s",
+                                     debugString(type).c_str()));
+}
+
+Element convert(Type type, double value) {
+  return convert(type, APFloat(value));
+}
+
+Element convert(Type type, std::complex<APFloat> value) {
+  if (isSupportedBooleanType(type))
+    return Element(type, value.real().isNonZero());
+  if (isSupportedIntegerType(type)) {
+    APSInt realVal(type.getIntOrFloatBitWidth(),
+                   /*isUnsigned=*/isSupportedUnsignedIntegerType(type));
+    bool roundingErr;
+    value.real().convertToInteger(realVal, APFloat::rmTowardZero, &roundingErr);
+    return Element(type, realVal);
+  }
+  if (isSupportedFloatType(type)) {
+    bool roundingErr;
+    auto realVal = value.real();
+    realVal.convert(type.cast<FloatType>().getFloatSemantics(),
+                    APFloat::rmNearestTiesToEven, &roundingErr);
+    return Element(type, realVal);
+  }
+  if (isSupportedComplexType(type)) {
+    const llvm::fltSemantics &fltSemantics = type.cast<ComplexType>()
+                                                 .getElementType()
+                                                 .cast<FloatType>()
+                                                 .getFloatSemantics();
+    bool roundingErr;
+    auto realVal = value.real();
+    realVal.convert(fltSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    auto imagVal = value.imag();
+    imagVal.convert(fltSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    return Element(type, std::complex<APFloat>(realVal, imagVal));
+  }
+  report_fatal_error(invalidArgument("Unsupported element type: %s",
+                                     debugString(type).c_str()));
+}
+
+Element convert(Type type, std::complex<double> value) {
+  return convert(type, std::complex<APFloat>(APFloat(value.real()),
+                                             APFloat(value.imag())));
+}
+
```

Comment:
Thank you for deduplicating the `convert` functions! However, there is still plenty of duplicated code remaining. Here's what I've come up with to minimize duplication: https://github.com/ghpvnist/stablehlo/pull/4.

Reply:
Thanks for the suggestions!